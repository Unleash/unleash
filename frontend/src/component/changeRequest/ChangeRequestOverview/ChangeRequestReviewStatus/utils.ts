export const getBrowserTimezoneInHumanReadableUTCOffset = (
    date = new Date(),
): string => {
    const offset = -date.getTimezoneOffset();
    const hours = Math.floor(Math.abs(offset) / 60);
    const minutes = Math.abs(offset) % 60;
    let sign = '+';
    if (offset < 0) {
        sign = '-';
    }

    // Ensure that hours and minutes are two digits
    const zeroPaddedHours = hours.toString().padStart(2, '0');
    const zeroPaddedMinutes = minutes.toString().padStart(2, '0');

    return `UTC${sign}${zeroPaddedHours}:${zeroPaddedMinutes}`;
};

export const getBrowserTimezone = (): string => {
    return Intl.DateTimeFormat().resolvedOptions().timeZone;
};
