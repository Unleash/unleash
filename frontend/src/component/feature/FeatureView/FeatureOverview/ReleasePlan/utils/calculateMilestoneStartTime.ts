import type { IReleasePlanMilestone } from 'interfaces/releasePlans';
import { addMinutes } from 'date-fns';

const parseStartTime = (startedAt: string | null | undefined): Date | null => {
    if (!startedAt) return null;

    const date = new Date(startedAt);
    if (Number.isNaN(date.getTime())) {
        return null;
    }
    return date;
};

const getIntervalMinutes = (
    milestone: IReleasePlanMilestone,
): number | null => {
    const intervalMinutes = milestone.transitionCondition?.intervalMinutes;

    if (!intervalMinutes) {
        return null;
    }
    return intervalMinutes;
};

const findMostRecentStartedMilestone = (
    milestones: IReleasePlanMilestone[],
    targetIndex: number,
): { index: number; startTime: Date } | null => {
    for (let i = targetIndex; i >= 0; i--) {
        const startTime = parseStartTime(milestones[i].startedAt);
        if (startTime) {
            return { index: i, startTime };
        }
    }
    return null;
};

const findBaselineMilestone = (
    milestones: IReleasePlanMilestone[],
    targetIndex: number,
    activeMilestoneId?: string,
): { index: number; startTime: Date } | null => {
    if (!activeMilestoneId) {
        return findMostRecentStartedMilestone(milestones, targetIndex);
    }

    const activeIndex = milestones.findIndex((m) => m.id === activeMilestoneId);
    if (activeIndex === -1 || activeIndex > targetIndex) {
        return findMostRecentStartedMilestone(milestones, targetIndex);
    }

    const activeStartTime = parseStartTime(milestones[activeIndex].startedAt);
    if (activeStartTime) {
        return { index: activeIndex, startTime: activeStartTime };
    }

    return findMostRecentStartedMilestone(milestones, targetIndex);
};

const calculateTimeFromBaseline = (
    milestones: IReleasePlanMilestone[],
    baseline: { index: number; startTime: Date },
    targetIndex: number,
): Date | null => {
    let currentTime = baseline.startTime;

    for (let i = baseline.index; i < targetIndex; i++) {
        const previousMilestone = milestones[i];
        const intervalMinutes = getIntervalMinutes(previousMilestone);

        if (!intervalMinutes) return null;

        currentTime = addMinutes(currentTime, intervalMinutes);
    }

    return currentTime;
};

export const calculateMilestoneStartTime = (
    milestones: IReleasePlanMilestone[],
    targetMilestoneId: string,
    activeMilestoneId?: string,
): Date | null => {
    const targetIndex = milestones.findIndex((m) => m.id === targetMilestoneId);
    if (targetIndex === -1) return null;

    const baseline = findBaselineMilestone(
        milestones,
        targetIndex,
        activeMilestoneId,
    );
    if (!baseline) return null;

    if (baseline.index === targetIndex) {
        return baseline.startTime;
    }

    return calculateTimeFromBaseline(milestones, baseline, targetIndex);
};
