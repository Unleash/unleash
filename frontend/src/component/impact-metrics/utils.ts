export const getTimeUnit = (selectedRange: string) => {
    switch (selectedRange) {
        case 'hour':
            return 'minute';
        case 'day':
            return 'hour';
        case 'week':
            return 'day';
        case 'month':
            return 'week';
        default:
            return 'day';
    }
};

export const getDisplayFormat = (selectedRange: string) => {
    switch (selectedRange) {
        case 'hour':
            return 'HH:mm';
        case 'day':
            return 'HH:mm';
        case 'week':
            return 'MMM dd';
        case 'month':
            return 'MMM dd';
        default:
            return 'MMM dd';
    }
};

export const getSeriesLabel = (metric: Record<string, string>): string => {
    const { __name__, ...labels } = metric;

    const labelParts = Object.entries(labels)
        .filter(([key, value]) => key !== '__name__' && value)
        .map(([key, value]) => `${key}=${value}`)
        .join(', ');

    if (!__name__ && !labelParts) {
        return 'Series';
    }

    if (!__name__) {
        return labelParts;
    }

    if (!labelParts) {
        return __name__;
    }

    return `${__name__} (${labelParts})`;
};

export const formatLargeNumbers = (value: number): string => {
    if (value >= 1000000) {
        return `${(value / 1000000).toFixed(0)}M`;
    }
    if (value >= 1000) {
        return `${(value / 1000).toFixed(0)}k`;
    }
    return value.toString();
};

export const getMetricType = (seriesName: string) => {
    if (seriesName.startsWith('unleash_counter_')) return 'counter';
    if (seriesName.startsWith('unleash_gauge_')) return 'gauge';
    return 'unknown';
};
