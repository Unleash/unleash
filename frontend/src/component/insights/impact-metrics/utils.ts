export const getTimeUnit = (selectedRange: string) => {
    switch (selectedRange) {
        case 'hour':
            return 'minute';
        case 'day':
            return 'hour';
        case 'week':
            return 'day';
        case 'month':
            return 'day';
        default:
            return 'hour';
    }
};

export const getDisplayFormat = (selectedRange: string) => {
    switch (selectedRange) {
        case 'hour':
        case 'day':
            return 'HH:mm';
        case 'week':
        case 'month':
            return 'MMM dd';
        default:
            return 'MMM dd HH:mm';
    }
};

export const getSeriesLabel = (metric: Record<string, string>): string => {
    const { __name__, ...labels } = metric;

    const labelParts = Object.entries(labels)
        .filter(([key, value]) => key !== '__name__' && value)
        .map(([key, value]) => `${key}=${value}`)
        .join(', ');

    if (!__name__ && !labelParts) {
        return 'Series';
    }

    if (!__name__) {
        return labelParts;
    }

    if (!labelParts) {
        return __name__;
    }

    return `${__name__} (${labelParts})`;
};

export const formatLargeNumbers = (value: number): string => {
    if (value >= 1000000) {
        return `${(value / 1000000).toFixed(0)}M`;
    }
    if (value >= 1000) {
        return `${(value / 1000).toFixed(0)}k`;
    }
    return value.toString();
};
