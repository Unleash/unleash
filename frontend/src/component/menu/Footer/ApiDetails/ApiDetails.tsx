import { ReactElement } from 'react';
import { ConditionallyRender } from 'component/common/ConditionallyRender/ConditionallyRender';
import {
    formatCurrentVersion,
    formatUpdateNotification,
    IPartialUiConfig,
} from './apidetails.helpers';
import { FooterTitle } from 'component/menu/Footer/FooterTitle';

interface IApiDetailsProps {
    uiConfig: IPartialUiConfig;
}

export const ApiDetails = (props: IApiDetailsProps): ReactElement => {
    const instanceId = props.uiConfig.versionInfo?.instanceId;
    const currentVersion = formatCurrentVersion(props.uiConfig);
    const environment = props.uiConfig.environment;
    const updateNotification = formatUpdateNotification(props.uiConfig);

    return (
        <section title="API details">
            <FooterTitle>
                {currentVersion}{' '}
                <ConditionallyRender
                    condition={Boolean(environment)}
                    show={<small>({environment})</small>}
                />
            </FooterTitle>
            <ConditionallyRender
                condition={Boolean(updateNotification)}
                show={
                    <small>
                        {updateNotification}
                        <br />
                    </small>
                }
            />
            <br />
            <small>{props.uiConfig.slogan}</small>
            <br />
            <ConditionallyRender
                condition={Boolean(instanceId)}
                show={<small>{`${instanceId}`}</small>}
            />
        </section>
    );
};
