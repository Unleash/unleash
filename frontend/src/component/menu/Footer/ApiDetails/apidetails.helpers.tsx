import { IVersionInfo } from 'interfaces/uiConfig';

export interface IPartialUiConfig {
    name: string;
    version: string;
    slogan?: string;
    environment?: string;
    versionInfo?: IVersionInfo;
}

export const formatCurrentVersion = (uiConfig: IPartialUiConfig): string => {
    const current = uiConfig.versionInfo?.current;

    if (current?.enterprise) {
        return `${uiConfig.name} ${current.enterprise}`;
    }

    if (current?.oss) {
        return `${uiConfig.name} ${current.oss}`;
    }

    return `${uiConfig.name} ${uiConfig.version}`;
};

export const formatUpdateNotification = (
    uiConfig: IPartialUiConfig
): string | undefined => {
    const latest = uiConfig.versionInfo?.latest;
    const isLatest = uiConfig.versionInfo?.isLatest;

    if (latest?.enterprise && !isLatest) {
        return `Upgrade available - Latest Enterprise release: ${latest.enterprise}`;
    }

    if (latest?.oss && !isLatest) {
        return `Upgrade available - Latest OSS release: ${latest.oss}`;
    }
};
