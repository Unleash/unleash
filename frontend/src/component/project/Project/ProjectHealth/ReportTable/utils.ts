import differenceInDays from 'date-fns/differenceInDays';
import { EXPERIMENT, OPERATIONAL, RELEASE } from 'constants/featureToggleTypes';

const FORTY_DAYS = 40;
const SEVEN_DAYS = 7;

export const toggleExpiryByTypeMap: Record<string, number> = {
    [EXPERIMENT]: FORTY_DAYS,
    [RELEASE]: FORTY_DAYS,
    [OPERATIONAL]: SEVEN_DAYS,
};

export const getDiffInDays = (date: Date, now: Date) => {
    return Math.abs(differenceInDays(date, now));
};

export const expired = (diff: number, type: string) => {
    return diff >= toggleExpiryByTypeMap[type];
};
