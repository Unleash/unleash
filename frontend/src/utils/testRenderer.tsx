import React, { FC } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render as rtlRender, RenderOptions } from '@testing-library/react';
import { SWRConfig } from 'swr';
import { ThemeProvider } from 'themes/ThemeProvider';
import { IPermission } from 'interfaces/user';
import { AnnouncerProvider } from 'component/common/Announcer/AnnouncerProvider/AnnouncerProvider';
import { AccessProviderMock } from 'component/providers/AccessProvider/AccessProviderMock';

export const render = (
    ui: JSX.Element,
    {
        route = '/',
        permissions = [],
        ...renderOptions
    }: { route?: string; permissions?: IPermission[] } & Omit<
        RenderOptions,
        'queries'
    > = {}
) => {
    window.history.pushState({}, 'Test page', route);

    const Wrapper: FC = ({ children }) => (
        <SWRConfig value={{ provider: () => new Map() }}>
            <AccessProviderMock permissions={permissions}>
                <ThemeProvider>
                    <AnnouncerProvider>
                        <BrowserRouter>{children}</BrowserRouter>
                    </AnnouncerProvider>
                </ThemeProvider>
            </AccessProviderMock>
        </SWRConfig>
    );

    return rtlRender(ui, {
        wrapper: Wrapper,
        ...renderOptions,
    });
};
