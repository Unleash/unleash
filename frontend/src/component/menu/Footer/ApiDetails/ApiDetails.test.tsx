import React from 'react';
import { ApiDetails } from 'component/menu/Footer/ApiDetails/ApiDetails';
import { render } from 'utils/testRenderer';

test('renders correctly with empty version', () => {
    const uiConfig = {
        name: 'Unleash',
        slogan: 'We are the best!',
        environment: 'test',
        version: '',
    };

    render(<ApiDetails uiConfig={uiConfig} />);
    expect(document.body).toMatchSnapshot();
});

test('renders correctly with ui-config', () => {
    const uiConfig = {
        name: 'Unleash',
        slogan: 'We are the best!',
        environment: 'test',
        version: '1.1.0',
    };

    render(<ApiDetails uiConfig={uiConfig} />);
    expect(document.body).toMatchSnapshot();
});

test('renders correctly without uiConfig', () => {
    const uiConfig = {
        name: 'Unleash',
        version: '1.1.0',
    };

    render(<ApiDetails uiConfig={uiConfig} />);
    expect(document.body).toMatchSnapshot();
});

test('renders correctly with versionInfo', () => {
    const uiConfig = {
        name: 'Unleash',
        version: '1.2.3',
        versionInfo: {
            instanceId: '1',
            isLatest: false,
            current: { enterprise: '1.2.3', oss: '1.2.3' },
            latest: { enterprise: '1.2.4', oss: '1.2.4' },
        },
    };

    render(<ApiDetails uiConfig={uiConfig} />);
    expect(document.body).toMatchSnapshot();
});
