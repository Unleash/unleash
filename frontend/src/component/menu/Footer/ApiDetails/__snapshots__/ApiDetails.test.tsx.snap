// Vitest Snapshot v1

exports[`renders correctly with empty version 1`] = `
<body>
  <div>
    <section
      title="API details"
    >
      <h2
        class="css-cxwipi-title"
      >
        Unleash 
         
        <small>
          (
          test
          )
        </small>
      </h2>
      <br />
      <small>
        We are the best!
      </small>
      <br />
    </section>
    <div
      aria-atomic="true"
      aria-live="polite"
      class="css-i8rqz1-container"
      data-testid="ANNOUNCER_ELEMENT_TEST_ID"
      role="status"
    />
  </div>
</body>
`;

exports[`renders correctly with ui-config 1`] = `
<body>
  <div>
    <section
      title="API details"
    >
      <h2
        class="css-cxwipi-title"
      >
        Unleash 1.1.0
         
        <small>
          (
          test
          )
        </small>
      </h2>
      <br />
      <small>
        We are the best!
      </small>
      <br />
    </section>
    <div
      aria-atomic="true"
      aria-live="polite"
      class="css-i8rqz1-container"
      data-testid="ANNOUNCER_ELEMENT_TEST_ID"
      role="status"
    />
  </div>
</body>
`;

exports[`renders correctly with versionInfo 1`] = `
<body>
  <div>
    <section
      title="API details"
    >
      <h2
        class="css-cxwipi-title"
      >
        Unleash 1.2.3
         
      </h2>
      <small>
        Upgrade available - Latest Enterprise release: 1.2.4
        <br />
      </small>
      <br />
      <small />
      <br />
      <small>
        1
      </small>
    </section>
    <div
      aria-atomic="true"
      aria-live="polite"
      class="css-i8rqz1-container"
      data-testid="ANNOUNCER_ELEMENT_TEST_ID"
      role="status"
    />
  </div>
</body>
`;

exports[`renders correctly without uiConfig 1`] = `
<body>
  <div>
    <section
      title="API details"
    >
      <h2
        class="css-cxwipi-title"
      >
        Unleash 1.1.0
         
      </h2>
      <br />
      <small />
      <br />
    </section>
    <div
      aria-atomic="true"
      aria-live="polite"
      class="css-i8rqz1-container"
      data-testid="ANNOUNCER_ELEMENT_TEST_ID"
      role="status"
    />
  </div>
</body>
`;
